tools/monitor/backend.py

```diff
@@ -0,0 +1,193 @@
+
+    def load_tasklist(self, mission_path, session):
+        tasks_file_path = os.path.join(mission_path, 'sessions', str(session), 'tasklist.py')  
+    
+        import telecommand as tc
+        import datetime
+        from radio.task_actions import Send, SendLoop, Sleep, Print, SendReceive, WaitMode
+        from devices import camera
+        from devices.adcs import AdcsMode
+        from devices.camera import CameraLocation, PhotoResolution
+        from devices.comm import BaudRate
+        
+        if not os.path.exists(tasks_file_path):
+            print("Could not find file {}".format(tasks_file_path))
+            return []
+
+        with open(tasks_file_path) as tasks_file:
+            exec(tasks_file)
+            tasks_file.close()
+
+        return tasks
+
+    def create_dictionary(self, tasklist):
+            commandsDict = OrderedDict()
+
+            for item in tasklist:
+                command = item[0]
+                if not isinstance(command, DownloadFile):
+                    continue
+                commandsDict[command._correlation_id] = command
+
+            return commandsDict
+
+    def generate_new_tasklist(self, commandsDict):
+        newTasks = []
+        for _, command in commandsDict.items():
+            newTasks.append(command)
+
+        return newTasks
+
+    def process_frame(self, frame, commandsDict, ui):
+        if not isinstance(frame, response_frames.common.FileSendSuccessFrame) and not isinstance(frame, response_frames.common.FileSendErrorFrame):
+            return
+
+        try:
+            tasklistCommandChunks = commandsDict[frame.correlation_id]._seqs
+            if frame._seq in tasklistCommandChunks:
+                tasklistCommandChunks.remove(frame._seq)
+            if len(tasklistCommandChunks) == 0:
+                del commandsDict[frame.correlation_id]
+        except KeyError:
+            pass   
+
+        stamp = datetime.datetime.now().time()
+        ui.logFrame(frame, stamp)
+        ui.update_tasklist(commandsDict)
+
+    def handle_commands(self, request, commandsDict, socket, ui):
+        command = json.loads(request)
+
+        if command[0] == 'M':
+            correlation_id = command[1]
+            try:
+                task = commandsDict[correlation_id]
+                task_string = json.dumps(task.__dict__)
+                ui.log("Sending task for file '{}'/{}, length {}.".format(task._path, task._correlation_id, len(task._seqs)))
+            except KeyError:
+                task_string = ""
+                ui.log("No data found for CID {}".format(correlation_id))
+            
+            socket.send(task_string)
+        elif command[0] == 'T':  
+            tasks = []
+            telemetryTasks = []
+            for _, command in commandsDict.items():
+                if "telemetry" in command._path:
+                    telemetryTasks.append(command)
+                else:
+                    tasks.append(command.__dict__)
```

Comment:
Czemu dla telemetrii nie wypakowujesz od razu dicta, tylko robisz to później?

Reply:
Sortowanie w nieco lepszy sposób.